### PR TITLE
test: add shared test path policy

### PR DESCRIPTION
**Pull request by akrock**
## Summary

Fixes #452.

**Test path policy**

- Adds `ForgeTrust.AppSurface.Testing` as a shared test-only helper project with `TestPathUtils.PathUnder`, `RelativePath`, and `FindRepoRoot`.
- Adds `ForgeTrust.AppSurface.Testing.Tests` with helper edge-case coverage and a repository-wide Roslyn policy test for risky dynamic fixture path construction.
- Migrates high-risk dynamic test fixture paths across Auth, Flow, Docs, Tailwind, RazorWire, MarkdownSnippets, PackageIndex, and Release tests to the shared helper.
- Removes duplicated local `TestPathUtils` copies and wires the shared helper into touched test projects with lockfile updates.

**Policy docs**

- Adds the helper README cookbook and `CONTRIBUTING.md` guidance.
- Adds a reasoned path-policy allowlist for intentional platform path behavior.
- Records the change in `CHANGELOG.md` and `releases/unreleased.md`.

## Test Coverage

CODE PATHS                                                  USER FLOWS
[+] test-only shared path helper                            [+] Test authors build fixture paths
  - [TESTED] PathUnder happy path                            - [TESTED] dynamic paths must use PathUnder
  - [TESTED] same-base "." path                              - [TESTED] rooted/traversal fixture paths rejected
  - [TESTED] traversal + sibling-prefix escape               - [TESTED] stale/invalid allowlist entries fail
  - [TESTED] null/blank/rooted/Windows-looking input         - [TESTED] docs explain migration and exceptions
  - [TESTED] symlink target non-resolution documented
  - [TESTED] repo-root discovery from file/dir/missing

[+] repository-wide fixture path policy
  - [TESTED] scanner flags dynamic Path.Join/Combine under base
  - [TESTED] ignores comments, strings, literals, non-test outputs
  - [TESTED] flags member, interpolated, concatenated, wrapper args
  - [TESTED] formatter gives replacement and allowlist guidance
  - [TESTED] allowlist requires reason and rejects expired/stale entries

Application code paths: none changed.

Coverage-solution result: line `95.70%` (`34989/36563`), branch `89.22%` (`11351/12722`).

## Pre-Landing Review

No issues found.

Production path inventory was run with `rg -n "Path\.(Combine|Join)\(" --glob "*.cs" -g "!**/*Tests*/**" -g "!tests/**" -g "!**/bin/**" -g "!**/obj/**"`. It found existing production `Path.Combine`/`Path.Join` usage, but this PR keeps production changes out of scope and did not introduce a new #452-linked production finding.

## Design Review

No frontend files changed; design review skipped.

## Eval Results

No prompt-related files changed; evals skipped.

## Scope Drift

Scope Check: CLEAN

Intent: implement a shared test-only path-under-base policy for fixture paths.

Delivered: shared helper, migrated risky test paths, policy enforcement test, docs guidance, release note, and verification.

## Plan Completion

- [DONE] Add shared `tests/ForgeTrust.AppSurface.Testing` helper with `TestPathUtils.PathUnder`, `RelativePath`, and `FindRepoRoot`.
- [DONE] Keep helper test-only and outside package surface.
- [DONE] Add helper test project, solution entries, and lockfiles.
- [DONE] Remove duplicated local helpers and migrate risky dynamic path construction.
- [DONE] Add Roslyn-based policy test and reasoned allowlist.
- [DONE] Add helper README cookbook and `CONTRIBUTING.md` guidance.
- [DONE] Verify package tooling does not classify the helper as package surface.
- [DONE] Run production path inventory and keep production changes out of scope.
- [DONE] Run build, solution tests, coverage script, and QA smoke evidence.

## Verification Results

- `dotnet restore ForgeTrust.AppSurface.slnx` passed.
- `dotnet build ForgeTrust.AppSurface.slnx --no-restore` passed with 0 warnings and 0 errors.
- `dotnet test ForgeTrust.AppSurface.slnx --no-build` passed on the final full rerun: 4512 tests, 0 failures.
- `dotnet test Web/ForgeTrust.RazorWire.IntegrationTests/ForgeTrust.RazorWire.IntegrationTests.csproj --no-build` passed: 99 tests, 0 failures.
- `./scripts/coverage-solution.sh` passed across 22 test projects; line `95.70%`, branch `89.22%`.
- `/qa` diff-aware browser smoke passed at `100/100`; report: `.gstack/qa-reports/qa-report-localhost-5189-2026-06-02.md`.

Note: the first full solution test run had one transient Playwright timeout in `RazorWireMvcPlaywrightTests.PublishMessage_BroadcastsToOtherSession`; the exact test rerun, the integration project rerun, the final full solution rerun, and the coverage run all passed.

## TODOS

No `TODOS.md` exists in this repo; no TODO items were updated.

## Documentation

Updated:

- `tests/ForgeTrust.AppSurface.Testing/README.md`
- `CONTRIBUTING.md`
- `CHANGELOG.md`
- `releases/unreleased.md`

## Test plan

- [x] Full solution build passes.
- [x] Full solution tests pass.
- [x] Coverage solution script passes.
- [x] Browser QA smoke passes.

### CODE REVIEW

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+            if (trimmed.StartsWith("- ", StringComparison.Ordinal))
+            {
+                AddEntry(entries, current);
+                current = new Dictionary<string, string>(StringComparer.Ordinal);
+                AddProperty(current, trimmed[2..]);
+                continue;
+            }
+
+            AddProperty(current, trimmed);
+        }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs]: Addressed in 50cae971 by applying the suggested allowlist parsing cleanup. I also updated the path-policy allowlist line key after the helper edit and reran the focused policy tests plus full solution/coverage gates.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs]: This duplicate thread is already addressed in 50cae971 and remains present at the current head: PathPolicyAllowlist.Parse now iterates content.Split('\n').Select(rawLine => rawLine.TrimEnd('\r')). Resolving the stale duplicate.

File: Web/ForgeTrust.AppSurface.Docs.Tests/RepositoryReadmePolicyTests.cs [hunk 3]
```diff
@@ -170,10 +170,14 @@ private static IReadOnlyList<PackageManifestReadmeEntry> ReadPackageManifestEntr
...
-
-        return Path.GetFullPath(normalizedPath, repoRoot);
+        try
+        {
+            return TestPathUtils.PathUnder(repoRoot, repositoryRelativePath);
+        }
+        catch (ArgumentException)
+        {
+            throw new InvalidOperationException($"{description} must be repository-relative and stay under the repository root.");
+        }
```
[the remark this answers is not in the document]
Comment (akrock) [reply Web/ForgeTrust.AppSurface.Docs.Tests/RepositoryReadmePolicyTests.cs]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: Auth/ForgeTrust.AppSurface.Auth.Tests/AppSurfaceAuthDependencyGuardTests.cs [hunk 4]
```diff
@@ -70,31 +70,14 @@ public void LoadProject_WithRootedPath_ThrowsArgumentException()
...
-            {
-                return current.FullName;
-            }
-
-            current = current.Parent;
+            throw new ArgumentException("Project path must be relative and stay under the repository root.", nameof(projectPath), exception);
         }
-
-        throw new InvalidOperationException("Could not find the AppSurface repository root.");
     }
```
[the remark this answers is not in the document]
Comment (akrock) [reply Auth/ForgeTrust.AppSurface.Auth.Tests/AppSurfaceAuthDependencyGuardTests.cs:74]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: Flow/ForgeTrust.AppSurface.Flow.DurableTask.Tests/DurableTaskDependencyGuardTests.cs [hunk 5]
```diff
@@ -33,23 +33,7 @@ private static IEnumerable<string> GetIncludes(XDocument project, string element
...
-                return current.FullName;
-            }
-
-            current = current.Parent;
-        }
-
-        throw new InvalidOperationException("Could not find repository root.");
+        var repositoryRoot = TestPathUtils.FindRepoRoot(AppContext.BaseDirectory);
+        return XDocument.Load(TestPathUtils.PathUnder(repositoryRoot, projectPath));
     }
```
[the remark this answers is not in the document]
Comment (akrock) [reply Flow/ForgeTrust.AppSurface.Flow.DurableTask.Tests/DurableTaskDependencyGuardTests.cs:48]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+    string Kind,
+    string RelativePath,
+    int Line,
+    int Column,
+    string Invocation,
+    string RiskyArgument,
+    string Reason)
+{
+    internal string Key => $"{Kind}|{RelativePath}|{Line}|{Column}|{Invocation}|{RiskyArgument}";
+}
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs:291]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+            .Select(path => Path.GetRelativePath(repositoryRoot, path).Replace('\\', '/'))
+            .Where(IsTestSourcePath)
+            .SelectMany(relativePath => ScanSource(
+                relativePath,
+                File.ReadAllText(TestPathUtils.PathUnder(repositoryRoot, relativePath))))
+            .OrderBy(violation => violation.RelativePath, StringComparer.Ordinal)
+            .ThenBy(violation => violation.Line)
+            .ThenBy(violation => violation.Column)
+            .ToArray();
+    }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs:335]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+
+    private PathPolicyAllowlist(ImmutableDictionary<string, Entry> entries)
+    {
+        _entries = entries;
+    }
+
+    internal static PathPolicyAllowlist Load(string path)
+    {
+        return Parse(File.ReadAllText(path));
+    }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs:515]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+
+        if (current.TryGetValue("expires", out var expiresText)
+            && DateOnly.TryParse(expiresText, out var expires)
+            && expires < DateOnly.FromDateTime(DateTime.UtcNow))
+        {
+            throw new InvalidOperationException($"Allowlist entry '{key}' expired on {expires:yyyy-MM-dd}.");
+        }
+
+        entries.Add(key, new Entry(category, reason));
+    }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+
+        foreach (var unmatchedKey in unmatchedKeys)
+        {
+            lines.Add(string.Empty);
+            lines.Add($"Unmatched allowlist entry: {unmatchedKey}");
+        }
+
+        return string.Join(Environment.NewLine, lines);
+    }
+}
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs:688]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestPathUtilsTests.cs [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+        Assert.Equal(repoRoot, TestPathUtils.FindRepoRoot(missingNestedDirectory));
+    }
+
+    [Fact]
+    public void FindRepoRoot_ShouldThrowWhenSolutionFileIsMissing()
+    {
+        Directory.CreateDirectory(_tempRoot);
+
+        Assert.Throws<InvalidOperationException>(() => TestPathUtils.FindRepoRoot(_tempRoot));
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestPathUtilsTests.cs:194]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing/TestPathUtils.cs [hunk 12]
```diff
@@ -83,12 +87,42 @@ public static string RelativePath(params string[] segments)
...
+            if (File.Exists(PathUnder(current.FullName, "ForgeTrust.AppSurface.slnx")))
+            {
+                return current.FullName;
+            }
+
+            current = current.Parent;
+        }
+
+        throw new InvalidOperationException("Could not locate repository root.");
+    }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing/TestPathUtils.cs:140]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: Web/ForgeTrust.AppSurface.Docs.Tests/AppSurfaceDocsHarvestVcsIgnorePolicyTests.cs [hunk 13]
```diff
@@ -804,17 +804,14 @@ private AppSurfaceDocsHarvestPathPolicySnapshot CreateSnapshot(Action<AppSurface
...
         }
-
-        var fullPath = Path.GetFullPath(Path.Join(_root, localRelativePath));
-        var rootPrefix = Path.TrimEndingDirectorySeparator(_root) + Path.DirectorySeparatorChar;
-        if (!fullPath.StartsWith(rootPrefix, StringComparison.Ordinal))
+        catch (ArgumentException exception)
         {
-            throw new ArgumentException("Fixture paths must stay under the test root.", nameof(relativePath));
+            throw new ArgumentException("Fixture paths must be relative and stay under the test root.", nameof(relativePath), exception);
         }
```
[the remark this answers is not in the document]
Comment (akrock) [reply Web/ForgeTrust.AppSurface.Docs.Tests/AppSurfaceDocsHarvestVcsIgnorePolicyTests.cs:826]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: Web/ForgeTrust.AppSurface.Docs.Tests/AppSurfaceDocsVersionArchiveControllerTests.cs [hunk 14]
```diff
@@ -414,7 +414,16 @@ private void PinReleaseManifestDigests(AppSurfaceDocsVersionCatalog catalog)
...
-            var candidatePath = Path.GetFullPath(Path.Join(basePath, version.ExactTreePath.Trim()));
+            string candidatePath;
+            try
+            {
+                candidatePath = TestPathUtils.PathUnder(basePath, version.ExactTreePath.Trim());
+            }
+            catch (ArgumentException)
+            {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (akrock) [reply Web/ForgeTrust.AppSurface.Docs.Tests/AppSurfaceDocsVersionArchiveControllerTests.cs:451]: Addressed in fb6af1c5: preserved helper exception context, normalized dependency guard path handling, added the requested guard/branch regression tests, documented FindRepoRoot and the path policy internal diagnostics/allowlist contracts, and reran focused tests plus solution build/tests.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 15]
```diff
@@ -0,0 +1,638 @@
...
+        {
+            throw new InvalidOperationException($"Allowlist entry '{key}' must include a reason (near line {lineNumber}).");
+        }
+
+        if (current.TryGetValue("expires", out var expiresText)
+            && DateOnly.TryParse(expiresText, out var expires)
+            && expires < DateOnly.FromDateTime(DateTime.UtcNow))
+        {
+            throw new InvalidOperationException($"Allowlist entry '{key}' expired on {expires:yyyy-MM-dd} (near line {lineNumber}).");
+        }
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs]: Addressed in 7475c81e: allowlist expires now uses DateOnly.TryParseExact with yyyy-MM-dd, CultureInfo.InvariantCulture, and DateTimeStyles.None; malformed expires values now fail validation, and a regression test covers the malformed date case.

File: tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs [hunk 16]
```diff
@@ -0,0 +1,638 @@
...
+        if (string.IsNullOrWhiteSpace(reason))
+        {
+            throw new InvalidOperationException($"Allowlist entry '{key}' must include a reason (near line {lineNumber}).");
+        }
+
+        if (current.TryGetValue("expires", out var expiresText)
+            && DateOnly.TryParse(expiresText, out var expires)
+            && expires < DateOnly.FromDateTime(DateTime.UtcNow))
+        {
+            throw new InvalidOperationException($"Allowlist entry '{key}' expired on {expires:yyyy-MM-dd} (near line {lineNumber}).");
```
[the remark this answers is not in the document]
Comment (akrock) [reply tests/ForgeTrust.AppSurface.Testing.Tests/TestFixturePathPolicyTests.cs]: Addressed in 7475c81e: allowlist expires now uses DateOnly.TryParseExact with yyyy-MM-dd, CultureInfo.InvariantCulture, and DateTimeStyles.None; malformed expires values now fail validation, and a regression test covers the malformed date case.
